File: api/src/main/java/tutmirleid/api/controller/ScenarioController.java
package tutmirleid.api.controller;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import tutmirleid.api.entity.ScenarioEntity;
import tutmirleid.api.model.Scenario;
import tutmirleid.api.service.ScenarioService;
import tutmirleid.api.utils.exception.ScenarioNotFoundException;

import java.util.List;

@RestController()
@RequestMapping(BaseController.API_URI + "/scenarios")
public class ScenarioController extends BaseController {
    private final ScenarioService scenarioService;

    public ScenarioController(ScenarioService scenarioService) {
        this.scenarioService = scenarioService;
    }

    @GetMapping()
    public ResponseEntity<Object> getScenarios() {
        try {
            List<Scenario> scenarios = scenarioService.getScenarios();

            return ResponseEntity.ok(scenarios);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Something went wrong...");
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<Object> getScenarioByID(@PathVariable Long id) {
        try {
            Scenario scenario = scenarioService.getScenarioWithSteps(id);

            return ResponseEntity.ok(scenario);
        } catch (ScenarioNotFoundException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Something went wrong...");
        }
    }

    @PostMapping("/create")
    public ResponseEntity<String> createScenario(@RequestBody ScenarioEntity scenarioEntity) throws Exception {
        try {
            if (scenarioEntity.getTitle().trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Scenario title can not be null");
            }

            this.scenarioService.createScenario(scenarioEntity);

            return ResponseEntity.ok("Scenario has been created successful!");
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Something went wrong...");
        }
    }

    @PatchMapping("/{id}")
    public ResponseEntity<Object> updateScenario(@PathVariable Long id,
                                                 @RequestBody ScenarioEntity scenarioEntity) {
        try {
            Scenario scenario = scenarioService.updateScenario(id, scenarioEntity);

            return ResponseEntity.ok(scenario);
        } catch (ScenarioNotFoundException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Something went wrong... " + e);
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Object> deleteScenario(@PathVariable Long id) {
        try {
            scenarioService.deleteScenario(id);

            return ResponseEntity.ok("Scenario has been deleted successful!");
        } catch (ScenarioNotFoundException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Something went wrong...");
        }
    }
}

File: src/scenario/scenario.controller.ts
import {Body, Controller, Delete, Get, Param, ParseIntPipe, Patch, Post} from "@nestjs/common";
import {ScenarioService} from "./scenario.service";
import {CreateScenarioDto} from "../dto/create.scenario.dto";
import {UpdateScenarioDto} from "../dto/update.scenario.dto";

@Controller('/scenario')
export class ScenarioController {
    constructor(private readonly scenarioService: ScenarioService) {}

    @Post()
    async createScenario(@Body() dto: CreateScenarioDto) {
        return await this.scenarioService.createScenario(dto);
    }

    @Get()
    async getScenarios() {
        return await this.scenarioService.getScenarios();
    }

    @Get(':id')
    async getScenario(@Param('id', ParseIntPipe) id: number) {
        return await this.scenarioService.getScenario(id);
    }

    @Patch(':id')
    async updateScenario(
        @Param('id') id: number,
        @Body() dto: UpdateScenarioDto
    ) {
        return await this.scenarioService.updateScenario(id, dto);
    }

    @Delete(':id')
    async deleteScenario(@Param('id') id: number) {
        return await this.scenarioService.deleteScenario(id);
    }
}